Modify doc index

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed

`doc/src/modules` is fixed.

1. integrals/index.rst and solvers/index.rst are added
2. Contents for submodule documentations are directly exposed to `modules/index/rst`.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

JSS95 commented:
@oscarbenjamin Before writing the document for `kind`, I cleaned up some document structures. After this, I'll write and add `kind.rst` to `modules/core` directory.

asmeurer commented on doc/src/tutorial/index.rst:
```diff
@@ -17,3 +17,4 @@
    solvers.rst
    matrices.rst
    manipulation.rst
+   numeric-computation.rst
```
This isn't supposed to go in the tutorial. The tutorial is an introductory tutorial. The numeric computation page is a guide (see https://documentation.divio.com/). I agree it doesn't blog in `modules`, which is the reference, but it doesn't belong here either. We should have a separate sections for how-to guides.

JSS95 replied:
What about `special-topics`, then?

JSS95 replied:
Then what about putting it into `special_topics`?

asmeurer replied:
That's also a good name. We just need some sort of top level directory for user guides so that people aren't tempted to put them in the tutorial (it happens too often).

JSS95 replied:
I renamed it to `how-to guides` and moved `numeric-computation.rst` to there.

asmeurer commented:
For GSoD last year, we decided that large pages like https://docs.sympy.org/latest/modules/core.html are fine, as long as there is an index (which there is, on the left panel). Although I think we may actually want to eventually move such pages to autosummary (see https://github.com/sympy/sympy/pull/18594).

oscarbenjamin commented:
I don't think it's good to have very long docs pages. The autosummary idea is much better but I'm not sure if that's orthogonal to this or not.

asmeurer commented:
Long docs pages have advantage. You can easily CMD-f on the page to find stuff. You can quickly scroll around to compare things. It's easy to see at a glance everything that is there. I think splitting pages often feels better, but in practice, a single long page actually is better. 

At the end of the day, this is something that it would be good to have some sort of user testing on. Or maybe other projects have already been studies of this that we can learn from.

JSS95 commented:
I reverted `core`.

akshatsood2249 commented:
How about this #20556 , if long pages are favourable then how about adding a side button for navigating to the top

sylee957 commented:
Autosummary works by generating many stub pages, which gets written in the source folder, so I don't think that it's very different than manually crafting new rst files for each classes, functions.

moorepants commented on doc/src/how_to_guides/index.rst:
```diff
@@ -0,0 +1,15 @@
+
+The purpose of this collection of documents is to provide users of SymPy with
+topics which are not strictly tutorial or are longer than tutorials and tests.
+The documents will hopefully show how SymPy can be used in more advanced topics.
```
I don't understand this explanation or reasoning for the new "how-to guides" section. We have a single sympy tutorial: https://docs.sympy.org/latest/tutorial/index.html (I think that is the only one), so I don't know what "tutorials" refers to.

There are other tutorial-like things, like `numeric-computation.rst`, or maybe you want to call them "how to guides". https://docs.sympy.org/latest/gotchas.html also seems like a "how-to guide".

We also have documentation generated from the `__doc__` attributes of modules, functions, classes, methods, etc. Some of the modules `__doc__` text is also very much like a tutorial/how-to guide (e.g. https://docs.sympy.org/latest/modules/printing.html).

We have things like this too: https://docs.sympy.org/latest/modules/physics/vector/vectors.html which are reference documents or maybe a how-to guide or something else.

I agree that we need some new organization wrt these types of docs, but I'm not sure adding a "How-To Guides" section is the best solution, especially when some of the other narrative docs are not included there.

The `__doc__` sections for classes, methods, functions, attributes now have to follow the SymPy docstring style guide. These can then clearly be separated out into "API documentation" (much like most other libraries do). The `__doc__` of many modules is a bit more confusing and some may belong in `API documentation` but some may not.

I've used the word "narrative" for describing the non-api docs in the past. Anything where you are writing sentences, paragraphs, mixed with code for explanation. Can we have only three headings in the top-level table of contents like:

```
Narrative
=======

   install.rst
   tutorial/index.rst
   gotchas.rst
   special_topics/index.rst
   documentation-style-guide.rst

API
===
   
modules/index.rst

Other
=====
   wiki.rst
   outreach.rst
   aboutus.rst
   citing.rst

```

The "other" section has things I think belong on the SymPy website, not necessarily in the docs. If so, we could remove that section and only have narrative and api sections.

Also, "narrative" is not a good name for the section header. Maybe "how-to guides" is better or something else. Or just "Guides".

The tricky thing is that there are still narrative documentation mixed into the modules/ folder. The https://docs.sympy.org/latest/modules/physics/vector/vectors.html is an example. It is unclear we we should extract those so they are separated from the API docs. I think most nice documentation websites do that.

I've written about this before in other email, github, etc. posts. I personally, would like to see the API and narrative documentation separated.

moorepants replied:
Pandas has 4 very clear divisions:

https://pandas.pydata.org/docs/

- Getting started
- User guide
- API
- Developer guide

That sounds nice to me. Maybe we could reorganize in that way?

JSS95 replied:
Perhaps, but it is too major to be done here. I reverted this change, but leaving this conversation open for future reference.

asmeurer replied:
@moorepants the distinction comes from https://documentation.divio.com/. I recommend watching the presentation on that page if you have the time. I believe NumPy also planning on using this division for their documentation. 

I agree "tutorials" plural in this text is a bit confusing. This is referring to the idea of tutorials in general, but we only have one tutorial for SymPy.

We absolutely need some sort of top-level section for guides. Whether it is called "how to guides" or something else I don't care. The problem is that such guides clearly don't belong in the API documentation. So when people want to add one, they look at what else is there and they see the tutorial. But the tutorial is NOT where these things should go either. The tutorial is meant to be an introductory tutorial for people new to SymPy, and these how-to guides are out of scope for it (again, please see the Divio presentation). But I've seen this happen a dozen times, including an earlier commit of this PR, where people want to add something that is essentially a how-to guide, and they try to put it in the tutorial, and I have to explain that it doesn't belong there. So we need some sort of obvious top level place in the docs where people will put these things that isn't the tutorial. I also hope having a clear place for these will encourage writing more long form non-API documentation, which is an area SymPy's documentation is weak in.

asmeurer replied:
Unfortunately @JSS95 force pushed so I can't read what was written here except for the blurb that GitHub shows for this comment. But based on what I can see here it looked fine to me, with maybe a few minor edits.

moorepants replied:
I checked out the divio thing. It seems reasonable, although I'm not sure that description covers everything you might want in documentation. Regardless, I'm fine adapting that for SymPy. It would be a significant improvement to the organization of our documentation.

JSS95 committed:
Fix doc/src/modules

Incorrect header depth of core.rst is fixed.
integrals/index.rst and solvers/index.rst are added

JSS95 commented:
Perhaps redesigning the doc structure needs more discussion in some other page. I reduced the scope of this PR to enhance the files in `src/modules` without any big change.

asmeurer commented on doc/src/modules/index.rst:
```diff
@@ -5,9 +5,12 @@ SymPy Modules Reference
+Every feature of SymPy have a test case. When you are not sure how to use
+something, just look into the example section in docstrings or ``tests/``
+directories. Find the feature and read the tests for it. It will tell you
+everything you need to know.
```
IMHO we shouldn't really be recommending people to use the tests as documentation. We should have proper examples and real docstrings.

JSS95 replied:
Fixed it.

asmeurer commented on doc/src/modules/index.rst:
```diff
@@ -29,24 +32,19 @@ access any SymPy module, or use the list below:
    matrices/index.rst
    ntheory.rst
    numeric-computation.rst
-   solvers/ode.rst
```
I'm -1 to these changes. Having these on the front page makes them easier to find.

asmeurer replied:
Unless the maxdepth makes them show up under solvers? Let me build the docs here and see what it looks like.

asmeurer replied:
No, it looks like it doesn't.

JSS95 replied:
I found that other documents such as `tensor`, `series`, `integrals` etc have same problem. I fixed them so that subdocs are exposed here by maxdepth. Please build again and check it!

asmeurer replied:
The removed sections still don't appear on the main index page.

JSS95 replied:
@asmeurer 
I increased the main index page toctree depth from 2 to 3. This will make subsections of other documents easier to find.

asmeurer commented:
Doing some of the other reorganization in another PR is fine. We also need to be careful that moving docs pages around will break any links on the web that point to the previous locations.

JSS95 committed:
Make subdocs of solvers visible in modules/index.rst

JSS95 committed:
Enhance tensor/index.rst

JSS95 committed:
Enhance series/index.rst

JSS95 committed:
Fix header levels and index of functions doc

JSS95 committed:
Enhance integrals/index.rst

moorepants commented:
This generally looks good to me. No objections.

JSS95 committed:
Fix typo

JSS95 commented:
@asmeurer Any more suggestion, please?

JSS95 committed:
Increase main index toctree depth from 2 to 3

oscarbenjamin commented:
This makes the index page very long. It now shows some things that it didn't before and in some cases that's good but in others I think it's quite distracting. Probably the best way would be to have a completely custom front page with (a small number of) sublinks to different broad sections of the docs more like what @moorepants suggested here: https://github.com/sympy/sympy/pull/20599#discussion_r543242808.

SymPy's module layout is highly imbalanced meaning that we have some modules like `sympy.algebras` that almost no one uses and then others like `sympy.solvers.ode` that are much bigger and more relevant for users. The index page here seems to suffer from the same problem where something like "preview" is on the same level as something like "differential equations".

JSS95 commented:
@oscarbenjamin Is it possible to apply different depth to each section in toctree? We need to expose the contents of `modules` to main index page.

JSS95 committed:
Apply different depths to toctree

JSS95 commented:
I modified the toctree so that different depths are applied to `modules` and other entries.

JSS95 committed:
Revert change in core.rst
